src/utils/page.js

```diff
@@ -1,29 +1,46 @@
 /**
- * @returns {boolean}
+ * @return {string}
  */
-export function isIframe () {
-  try {
-    return window.self !== window.top
-  } catch (e) {
-    return true
-  }
+export function loadedDomain () {
+  return (document.domain || (document.location && document.location.host)) || (window && window.location && window.location.host) || 'localhost'
 }
 
 /**
- * @returns {string}
- * @private
+ * @return {string|undefined}
  */
-export function getPage () {
-  try {
-    return isIframe() ? document.referrer : document.location.href
-  } catch (e) {
-    return document.location.href
-  }
+export function getReferrer (win = window) {
+  return _safeGet(() => win.top.document.referrer)
 }
 
 /**
- * @return {string}
+ * @return {string|undefined}
  */
-export function loadedDomain () {
-  return (document.domain || (document.location && document.location.host)) || (window && window.location && window.location.host) || 'localhost'
+export function getPage (win = window) {
+  const ancestorOrigins = _safeGet(() => win.location.ancestorOrigins) || {}
+
+  const windows = []
+  let currentWindow
+  do {
+    currentWindow = currentWindow ? currentWindow.parent : win
+    windows.push(currentWindow)
+  } while (currentWindow !== top)
```

Comment:
Might be a bit shorter, if you prefer `for`
```suggestion
  const windows = [win]
  
  for (let currentWindow = win; currentWindow !== top; currentWindow = currentWindow.parent) {
    windows.push(currentWindow)
  }
```

Reply:
The semantic is different. For loop doesn't keep the window #0, if it's already "top". My code does.

Reply:
`const windows = [win]` I didn't manage to come up with this and my for loop didn't work :facepalm: 

:+1:

Reply:
this code is wrong